lib/app/sequence_base.rb

```diff
@@ -717,28 +717,31 @@ def validate_reference_resolutions(resource, resolved_references = Set.new, max_
             next
           end
 
-          begin
-            # Should potentially update valid? method in fhir_dstu2_models
-            # to check for this type of thing
-            # e.g. "patient/54520" is invalid (fhir_client resource_class method would expect "Patient/54520")
-            if value.relative?
-              begin
-                value.resource_class
-              rescue NameError
-                problems << "#{path} has invalid resource type in reference: #{value.type}"
-                next
-              end
+          # Should potentially update valid? method in fhir_dstu2_models
+          # to check for this type of thing
+          # e.g. "patient/54520" is invalid (fhir_client resource_class method would expect "Patient/54520")
+          if value.relative?
+            begin
+              value.resource_class
+            rescue NameError
+              problems << "#{path} has invalid resource type in reference: #{value.type}"
+              next
             end
-            reference = value.reference
-            reference_type = value.resource_type
-            resolved_resource = value.read
+          end
+          reference = value.reference
+          reference_type = value.resource_type
 
-            raise InvalidReferenceResource if resolved_resource&.resourceType != reference_type
+          begin
+            resolved_resource = value.read
+          rescue ClientException => e
+            # Skip reference read error is the resource is not a US Core resource.
+            problems << "#{path} did not resolve: #{e}" if e.reply.response[:code] != 401 || Inferno::ValidationUtil::RESOURCES[:r4].key?(reference_type)
```

Comment:
The logic here is Inferno will report error if
(error code is not 401) OR (error code is 401 AND the referenced resource is a US Core resource)

During the SMART App Launch test, Inferno only asks for permission to read all US Core resources. So at here, if server reject reading reference to a US Core resource, this section causes that and raise an error. If server reject reading reference to a Non US Core resource, we think that is ok because Inferno does not request that permission. 

There is one minor gap. Though Inferno does not request permission for Non US Core resources, server MAY still grant it. In such scenario, if server reject reading a reference to a granted Non USCore resource, this section does NOT catch that. I think this is a problem but beyond Inferno's testing scope.

Reply:
Is there a reason we're only checking for 401? There are other possible response statuses: https://www.hl7.org/fhir/security.html#AccessDenied

Reply:
Is `404` allowed in this case?  Instead of going down this rabbit hole, should we just allow it to "not work", and not get to particular about how it doesn't work?

Reply:
Hmmm. We allows both `401` and `403` in [USCore311ONCAccessVerifyRestrictedSequence](https://github.com/onc-healthit/inferno-program/blob/4ce8b77dd6b328c0268c33414936b1593dc4d0fb/lib/modules/uscore_v3.1.1/access_verify_restricted_sequence.rb#L77). I will update this section (and the fhirUser section) to match that.